auth/context.go

```diff
@@ -0,0 +1,65 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import "context"
+
+var (
+	rawKey     = rawType{}
+	subjectKey = subjectType{}
+	groupsKey  = groupsType{}
+)
+
+type rawType struct{}
+type subjectType struct{}
+type groupsType struct{}
```

Comment:
Is it possible to have all three of them in the same time? Can we actually have one key with an enum type?

Reply:
Maybe have a "AuthContext" struct which embeds different auth types?

Reply:
Is there a library that provides this?

Reply:
> Is it possible to have all three of them in the same time? Can we actually have one key with an enum type?

Yes, and quite common actually.

> Maybe have a "AuthContext" struct which embeds different auth types?

Those are not auth types. The raw auth data might be a JWT, which contains the principal and group memberships. The authenticator (OIDC, for instance) will then place all the three components in the context. Another authenticator might get the subject and groups from the auth server.

I can split this if we need more time, but I think the third commit brings only a minimal API, generic enough to be extended with custom types in the future. The idea about AuthContext has been brought by @tigrannajaryan as well, though:

https://github.com/open-telemetry/opentelemetry-collector/commit/6b6e713310ef1296b417ff144f343b621d2aaf94#diff-208f8af800c6aa88673523930b718a5d1a5b3f2a2d2f6edbdcf97171ce3d15ca